ghcide/test/exe/ExceptionTests.hs

```diff
@@ -0,0 +1,212 @@
+              _ -> liftIO $ assertFailure $ show lens
+      , testCase "InvalidParams over InvalidUserState" $ do
+        let pluginId = "invalid-params-order"
+            plugins = pluginDescToIdePlugins $
+                [ (defaultPluginDescriptor pluginId)
+                    { pluginHandlers = mconcat
+                        [ mkPluginHandler SMethod_TextDocumentCodeLens $ \_ _ _-> do
+                            throwError $ PluginInvalidParams "error test"
+                         ,mkPluginHandler SMethod_TextDocumentCodeLens $ \_ _ _-> do
+                            throwError $ PluginInvalidUserState "error test"
+                        ]
+                    }
+                  , Notifications.descriptor (cmapWithPrio LogNotifications recorder) "ghcide-core"]
+
+        testIde recorder (testingLite recorder logger plugins) $ do
+            doc <- createDoc "A.hs" "haskell" "module A where"
+            waitForProgressDone
+            (view L.result -> lens) <- request SMethod_TextDocumentCodeLens (CodeLensParams Nothing Nothing doc)
+            case lens of
+              Left (ResponseError {_code = InR ErrorCodes_InvalidParams, _message}) ->
+                  liftIO $ assertBool "We caught an error, but it wasn't ours!"
+                          (T.isInfixOf "error test" _message && T.isInfixOf (coerce pluginId) _message)
+              _ -> liftIO $ assertFailure $ show lens
+      , testCase "InvalidUserState over RequestRefused" $ do
+        let pluginId = "invalid-user-state-order"
+            plugins = pluginDescToIdePlugins $
+                [ (defaultPluginDescriptor pluginId)
+                    { pluginHandlers = mconcat
+                        [ mkPluginHandler SMethod_TextDocumentCodeLens $ \_ _ _-> do
+                            throwError $ PluginInvalidUserState "error test"
+                         ,mkPluginHandler SMethod_TextDocumentCodeLens $ \_ _ _-> do
+                            throwError $ PluginRequestRefused "error test"
+                        ]
+                    }
+                  , Notifications.descriptor (cmapWithPrio LogNotifications recorder) "ghcide-core"]
+
+        testIde recorder (testingLite recorder logger plugins) $ do
+            doc <- createDoc "A.hs" "haskell" "module A where"
+            waitForProgressDone
+            (view L.result -> lens) <- request SMethod_TextDocumentCodeLens (CodeLensParams Nothing Nothing doc)
+            case lens of
+              Left (ResponseError {_code = InL LSPErrorCodes_RequestFailed, _message}) ->
+                  liftIO $ assertBool "We caught an error, but it wasn't ours!"
+                          (T.isInfixOf "error test" _message && T.isInfixOf (coerce pluginId) _message)
+              _ -> liftIO $ assertFailure $ show lens
+     ]]
+
+testIde :: Recorder (WithPriority Log) -> IDE.Arguments -> Session () -> IO ()
```

Comment:
Are we duplicating stuff from `func-tests` or something? I'm surprised the functions we need aren't already defined?

Reply:
This was shared with unit tests, so I moved the function from both there and unit tests and moved it to testutils